src/anyio/_core/_tempfile.py

```diff
@@ -0,0 +1,454 @@
+    :param suffix: The suffix for the temporary file name.
+    :param prefix: The prefix for the temporary file name.
+    :param dir: The directory in which the temporary file is created.
+    :param errors: The error handling scheme used for encoding/decoding errors.
+    """
+
+    def __init__(
+        self,
+        max_size: int = 0,
+        mode: str = "w+b",
+        buffering: int = -1,
+        encoding: str | None = None,
+        newline: str | None = None,
+        suffix: AnyStr | None = None,
+        prefix: AnyStr | None = None,
+        dir: AnyStr | None = None,
+        *,
+        errors: str | None = None,
+    ) -> None:
+        self.max_size = max_size
+        self.mode = mode
+        self.buffering = buffering
+        self.encoding = encoding
+        self.newline = newline
+        self.suffix: AnyStr | None = suffix
+        self.prefix: AnyStr | None = prefix
+        self.dir: AnyStr | None = dir
+        self.errors = errors
+        self._async_file: AsyncFile | None = None
+        self._fp: Any = None
+
+    async def __aenter__(
+        self: SpooledTemporaryFile[AnyStr],
+    ) -> SpooledTemporaryFile[AnyStr]:
+        fp = await to_thread.run_sync(
+            partial(
+                tempfile.SpooledTemporaryFile,
+                self.max_size,
+                self.mode,
+                self.buffering,
+                self.encoding,
+                self.newline,
+                self.suffix,
+                self.prefix,
+                self.dir,
+                errors=self.errors,
+            )
+        )
+        self._fp = fp
+        self._async_file = AsyncFile(fp)
+        return self
+
+    async def rollover(self) -> None:
+        if self._fp is None:
+            raise RuntimeError("Underlying file is not initialized.")
+
+        await to_thread.run_sync(cast(Callable[[], None], self._fp.rollover))
+
+    @property
+    def closed(self) -> bool:
+        if self._fp is None:
+            return True
+
+        return bool(self._fp.closed)
+
+    def __getattr__(self, attr: str) -> Any:
+        if self._async_file is not None:
+            return getattr(self._async_file, attr)
+
+        raise AttributeError(f"{self.__class__.__name__} has no attribute {attr}")
+
+    async def write(self, s: Any) -> int:
+        """
+        Asynchronously write data to the spooled temporary file.
+
+        If the file has not yet been rolled over, the data is written synchronously,
+        and a rollover is triggered if the size exceeds the maximum size.
+
+        :param s: The data to write.
+        :return: The number of bytes written.
+        :raises RuntimeError: If the underlying file is not initialized.
+        """
+        if self._fp is None:
+            raise RuntimeError("Underlying file is not initialized.")
+
+        if not getattr(self._fp, "_rolled", True):
+            result = self._fp.write(s)
```

Comment:
We need to have a bit of a think here. According to the rules of trio every async function should always checkpoint, however for SpooledTemporyFile to be useful it should be as close to a BytesIO in terms of performance at small sizes. The documentation should have a big red note here or write should include a checkpoint if we're writing to the memory buffer